src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -23432,6 +23566,28 @@ static void F(A<object?>[] c)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "a2.F").WithLocation(15, 13));
         }
 
+        [Fact]
+        public void SelectAnonymousType()
+        {
+            var source =
+@"using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int? E;
+    static void F(IEnumerable<C> c)
+    {
+        const int F = 0;
+        c.Select(o => new { E = o.E ?? F });
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (5,10): warning CS0649: Field 'C.E' is never assigned to, and will always have its default value 
+                //     int? E;
+                Diagnostic(ErrorCode.WRN_UnassignedInternalField, "E").WithArguments("C.E", "").WithLocation(5, 10));
+        }
```

Comment:
Not related to this PR: I feel like I asked before, but I couldn't remember where, or the outcome. What are the nullability rules for anonymous types? I assume it's similar to tuples (ie. it's just a type with members that are both `in` and `out` so we track the members and prevent any assignments from nullable to non-nullable and vice-versa).

Reply:
Hum, for anonymous types, isn't there a problem even if the fields are readonly?
`var a = condition ? one_anon_type_without_nullable ? same_anon_type_with_nullable;`

Reply:
We are reporting nullability mismatches for anonymous type fields. Added test for several more cases in `AnonymousTypes_05`.

---
In reply to: [192810121](https://github.com/dotnet/roslyn/pull/27283#discussion_r192810121) [](ancestors = 192810121)